.github/workflows/contracts-build-test.yml

```diff
@@ -0,0 +1,135 @@
+name: TON - Contracts Build and Test
+
+permissions:
+  contents: read
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+  workflow_dispatch:
+    inputs:
+      create_release:
+        description: "Publish a GitHub release on this manual run?"
+        type: boolean
+        default: false
+        required: false
+      sha:
+        description: "Commit SHA to build/release (optional)"
+        type: string
+        required: false
+
+jobs:
+  changes:
+    name: Detect changes
+    runs-on: ubuntu-latest
+    outputs:
+      ton_changes: ${{ steps.changes.outputs.ton_changes }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+      - name: Detect changes
+        id: changes
+        uses: dorny/paths-filter@de90cc6fb38fc0963ad72b210f1f284cd68cea36 # v3.0.2
+        with:
+          list-files: "shell"
+          filters: |
+            ton_changes:
+              - 'contracts/**'
+              - '.github/workflows/contracts-build-test.yml'
+
+  build:
+    name: Build and test contracts
+    needs: [ changes ]
+    # Run if changes are detected OR we're manually dispatching (for testing/release)
+    if: ${{ needs.changes.outputs.ton_changes == 'true' || github.event_name == 'workflow_dispatch' }}
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+    steps:
+      - name: Check out code
+        uses: actions/checkout@0ad4b8fadaa221de15dcec353f45205ec38ea70b # v4.1.4
+        with:
+          fetch-depth: 0
+
+      - name: Get Long and Short SHAs
+        id: get_sha
+        run: |
+          set -euo pipefail
+          TARGET_SHA="${{ github.event_name == 'workflow_dispatch' && github.event.inputs.sha != '' && github.event.inputs.sha || github.sha }}"
+          FULL_SHA="$(git rev-parse "${TARGET_SHA}")"
+          {
+            echo "short_sha=${FULL_SHA:0:12}"
+            echo "full_sha=${FULL_SHA}"
+          } >> "$GITHUB_OUTPUT"
+
+      - name: Install Nix
+        uses: cachix/install-nix-action@02a151ada4993995686f9ed4f1be7cfbb229e56f # v31
+        with:
+          nix_path: nixpkgs=channel:nixos-unstable
+
+      - name: Build (yarn)
+        run: |
+          set -euo pipefail
+          pushd contracts
+          nix develop .#contracts -c yarn && yarn build
+          popd
+
+      - name: Run tests
+        run: |
+          set -euo pipefail
+          pushd contracts
+          nix develop .#contracts -c  yarn test
+          popd
```

Comment:
No need to test contract after a PR is merged to main - assume all checks are green at that point